File: packages/simulation/tests/integration/golden-fixtures.test.ts
/**
 * Golden Fixtures Test
 *
 * Tests TypeScript simulator against golden fixtures.
 * Ensures simulator produces expected results on canonical test cases.
 */

import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { simulateFromInput } from '../../../src/core/contract-adapter';
import { SimInputSchema, SimResultSchema } from '../../../src/types/contracts';
import type { SimInput, SimResult } from '../../../src/types/contracts';

// CommonJS __dirname equivalent
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const fixturesDir = join(__dirname, '../../../../tools/telegram/tests/fixtures');

describe('Golden Fixtures', () => {
  const inputs = JSON.parse(readFileSync(join(fixturesDir, 'golden_sim_inputs.json'), 'utf8'));
  const expected = JSON.parse(readFileSync(join(fixturesDir, 'golden_sim_results.json'), 'utf8'));

  // Create a map of expected results by name
  const expectedMap = new Map(
    expected.map((e: { name: string; expected_result: SimResult }) => [e.name, e.expected_result])
  );

  it.each(inputs.map((input: SimInput & { name: string }, i: number) => [input.name, input]))(
    'should match golden fixture: %s',
    async (name: string, inputData: SimInput & { name: string }) => {
      // Remove name from input
      const { name: _, ...input } = inputData;

      // Validate and parse input
      const input_parsed = SimInputSchema.parse(input);

      // Run simulation
      const result = await simulateFromInput(input_parsed);

      // Validate result schema
      const validatedResult = SimResultSchema.parse(result);

      // Check if we have expected result for this fixture
      const expectedResult = expectedMap.get(name);
      if (expectedResult) {
        // Compare with expected (within tolerance)
        expect(validatedResult.final_pnl).toBeCloseTo(expectedResult.final_pnl, 2);
        expect(validatedResult.events.length).toBeGreaterThanOrEqual(expectedResult.events.length);
        expect(validatedResult.entry_price).toBeCloseTo(expectedResult.entry_price, 2);
        expect(validatedResult.final_price).toBeCloseTo(expectedResult.final_price, 2);
      } else {
        // No expected result, just verify it's valid
        expect(validatedResult.run_id).toBe(input_parsed.run_id);
        expect(validatedResult.total_candles).toBe(input_parsed.candles.length);
      }
    }
  );
});

File: packages/ingestion/python/telegram/tests/integration/golden_fixtures_test.py
"""
Golden Fixtures Test

Tests Python simulator against golden fixtures.
Ensures simulator produces expected results on canonical test cases.
"""

import pytest
import json
import duckdb
from pathlib import Path
from simulation.contracts import SimInput, SimResult
from simulation.simulator import DuckDBSimulator

FIXTURES_DIR = Path(__file__).parent.parent / 'fixtures'


@pytest.fixture
def golden_inputs():
    """Load golden input fixtures"""
    with open(FIXTURES_DIR / 'golden_sim_inputs.json') as f:
        return json.load(f)


@pytest.fixture
def golden_results():
    """Load golden result fixtures"""
    with open(FIXTURES_DIR / 'golden_sim_results.json') as f:
        return json.load(f)


@pytest.mark.integration
def test_golden_fixtures(golden_inputs, golden_results):
    """Test Python sim against golden fixtures"""
    con = duckdb.connect(':memory:')
    simulator = DuckDBSimulator(con)
    
    # Create expected map
    expected_map = {e['name']: e['expected_result'] for e in golden_results}
    
    for input_data in golden_inputs:
        name = input_data.pop('name')
        sim_input = SimInput.from_dict(input_data)
        result = simulator.run_from_contract(sim_input)
        
        # Validate results match (within tolerance)
        expected = expected_map.get(name)
        if expected:
            expected_result = SimResult.from_dict(expected)
            
            # Compare final PnL (within 2 decimal places)
            assert abs(result.final_pnl - expected_result.final_pnl) < 0.01, \
                f"{name}: final_pnl mismatch: {result.final_pnl} vs {expected_result.final_pnl}"
            
            # Compare entry/final prices
            assert abs(result.entry_price - expected_result.entry_price) < 0.01, \
                f"{name}: entry_price mismatch"
            assert abs(result.final_price - expected_result.final_price) < 0.01, \
                f"{name}: final_price mismatch"
            
            # Compare event counts (should be close)
            assert abs(len(result.events) - len(expected_result.events)) <= 1, \
                f"{name}: event count mismatch: {len(result.events)} vs {len(expected_result.events)}"
        else:
            # No expected result, just verify it's valid
            assert result.run_id == sim_input.run_id
            assert result.total_candles == len(sim_input.candles)
            assert result.final_pnl >= 0
    
    con.close()

